# **Perancangan Data Warehouse dan Analisis OLAP pada Data Manajemen Staf Rumah Sakit Menggunakan PostgreSQL dan Atoti**
### Kelompok 8 (2024 E)
Ananda Keissa Az Zahra (24031554051), Laili Nurrohmatul Fadhila Z. (24031554093), Eka Putri Maharani (24031554121)

**Dataset:** https://catalog.data.gov/dataset/hospital-staffing-2009-2013

## 0. Install Dependencies

In [1]:
!pip install psycopg2-binary sqlalchemy atoti pandas numpy apscheduler -q


[notice] A new release of pip is available: 26.0.1 -> 26.1.2
[notice] To update, run: python.exe -m pip install --upgrade pip


## 1. Download Dataset

In [2]:
import requests
import io
import os

DATASET_URL = "https://data.chhs.ca.gov/dataset/2417ea23-4b40-4953-b98d-6e26bc895a70/resource/4f36c327-91f8-4ae1-8044-e11d52808593/download/hospital-staffing-2009-2013.csv"
DATASET_PATH = 'hospital-staffing-2009-2013-.csv'

print(f"URL: {DATASET_URL}")

try:
    response = requests.get(DATASET_URL, timeout=60)
    response.raise_for_status()

    with open(DATASET_PATH, 'wb') as f:
        f.write(response.content)
    
    print(f"Status HTTP   : {response.status_code}")
    print(f"Ukuran file   : {len(response.content) / 1024:.1f} KB")
    print(f"File tersimpan: {DATASET_PATH}")
    
except requests.exceptions.ConnectionError:
    if os.path.exists(DATASET_PATH):
        print("Koneksi gagal, menggunakan file lokal yang sudah ada.")
    else:
        raise FileNotFoundError(f"File {DATASET_PATH} tidak ditemukan dan URL tidak dapat diakses.")


URL: https://data.chhs.ca.gov/dataset/2417ea23-4b40-4953-b98d-6e26bc895a70/resource/4f36c327-91f8-4ae1-8044-e11d52808593/download/hospital-staffing-2009-2013.csv
Koneksi gagal, menggunakan file lokal yang sudah ada.


## 2. Periodic Extraction

In [3]:
import pandas as pd
import numpy as np
import time
import asyncio
from datetime import datetime
from apscheduler.schedulers.background import BackgroundScheduler

DATASET_PATH = 'hospital-staffing-2009-2013-.csv'

df_raw_global = None

def extract_job():
    global df_raw_global
    timestamp = datetime.now().strftime('%Y-%m-%d %H:%M:%S')
    
    df = pd.read_csv(DATASET_PATH)
    df_raw_global = df
    return df

df_raw_global = extract_job()

scheduler = BackgroundScheduler()
scheduler.add_job(extract_job, 'interval', seconds=60, id='etl_job')
scheduler.start()

df_raw = df_raw_global.copy()
print(f"\nTotal baris ter-ekstrak: {len(df_raw)}")
print(f"Kolom: {list(df_raw.columns)}")


Total baris ter-ekstrak: 37604
Kolom: ['Year', 'Facility Number', 'Facility Name', 'Begin Date', 'End Date', 'County Name', 'Type of Control', 'Hours Type', 'Productive Hours', 'Productive Hours per Adjusted Patient Day']


## 3. ETL — Extract 

In [4]:
print(f"Total baris  : {len(df_raw)}")
print(f"Total kolom  : {len(df_raw.columns)}")
display(df_raw.head(3))

df_raw.info()

print(df_raw.isnull().sum())

Total baris  : 37604
Total kolom  : 10


,Year,Facility Number,Facility Name,Begin Date,End Date,County Name,Type of Control,Hours Type,Productive Hours,Productive Hours per Adjusted Patient Day
0,2009,106010735.0,ALAMEDA HOSPITAL,07/01/2008,06/30/2009,Alameda,District,Management & Supervision,63558,1.17
1,2009,106010735.0,ALAMEDA HOSPITAL,07/01/2008,06/30/2009,Alameda,District,Technician & Specialist,163706,3.02
2,2009,106010735.0,ALAMEDA HOSPITAL,07/01/2008,06/30/2009,Alameda,District,Registered Nurse,180034,3.32


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 37604 entries, 0 to 37603
Data columns (total 10 columns):
 #   Column                                     Non-Null Count  Dtype  
---  ------                                     --------------  -----  
 0   Year                                       37604 non-null  int64  
 1   Facility Number                            37519 non-null  float64
 2   Facility Name                              37519 non-null  object 
 3   Begin Date                                 37519 non-null  object 
 4   End Date                                   37519 non-null  object 
 5   County Name                                37604 non-null  object 
 6   Type of Control                            37519 non-null  object 
 7   Hours Type                                 37604 non-null  object 
 8   Productive Hours                           37604 non-null  int64  
 9   Productive Hours per Adjusted Patient Day  37417 non-null  float64
dtypes: float64(2), int64(2

---
## 4. ETL — Transform


### 4.1 Handle Missing Values + Bad Records Table (Vaisman)


In [5]:
from datetime import datetime

bad_records_list = []

def log_bad_records(df_bad, reason):
    if len(df_bad) == 0:
        return
    temp = df_bad.copy()
    temp['error_reason']    = reason
    temp['error_timestamp'] = datetime.now().strftime('%Y-%m-%d %H:%M:%S')
    bad_records_list.append(temp)
    print(f"[BAD RECORDS] {len(df_bad)} baris → '{reason}'")

df_work = df_raw_global.copy()
before  = len(df_work)

kolom_kritis = ['Productive Hours', 'Facility Number', 'Year', 'Begin Date', 'End Date']
mask_missing = df_work[kolom_kritis].isnull().any(axis=1)
log_bad_records(df_work[mask_missing], 'MISSING_CRITICAL_FIELD')
df_work = df_work[~mask_missing].copy()

mask_neg = df_work['Productive Hours'] < 0
log_bad_records(df_work[mask_neg], 'NEGATIVE_PRODUCTIVE_HOURS')
df_work = df_work[~mask_neg].copy()

mask_dup = df_work.duplicated(keep='first')
log_bad_records(df_work[mask_dup], 'DUPLICATE_ROW')
df_work = df_work[~mask_dup].copy()

if bad_records_list:
    df_bad_records = pd.concat(bad_records_list, ignore_index=True)
else:
    df_bad_records = pd.DataFrame(columns=list(df_raw_global.columns) + ['error_reason', 'error_timestamp'])

df_clean = df_work.copy()

print(f"Baris input      : {before:,}")
print(f"Total bad records: {len(df_bad_records)}")
print(f"Baris bersih     : {len(df_clean):,}")
if len(df_bad_records) > 0:
    display(df_bad_records.groupby('error_reason').size().reset_index(name='jumlah'))
else:
    print("  Tidak ada bad records.")


[BAD RECORDS] 85 baris → 'MISSING_CRITICAL_FIELD'
Baris input      : 37,604
Total bad records: 85
Baris bersih     : 37,519


,error_reason,jumlah
0,MISSING_CRITICAL_FIELD,85


### 4.2 Type Conversion & Standardization


In [6]:
df_clean['Begin Date'] = pd.to_datetime(df_clean['Begin Date'], errors='coerce')
df_clean['End Date']   = pd.to_datetime(df_clean['End Date'],   errors='coerce')

mask_bad_date = df_clean['Begin Date'].isna() | df_clean['End Date'].isna()
log_bad_records(df_clean[mask_bad_date], 'INVALID_DATE_FORMAT')
df_clean = df_clean[~mask_bad_date].copy()

df_clean['Reporting Duration Days'] = (df_clean['End Date'] - df_clean['Begin Date']).dt.days
df_clean['Quarter'] = df_clean['Begin Date'].dt.quarter
df_clean['Month']   = df_clean['Begin Date'].dt.month

df_clean['Facility Number'] = df_clean['Facility Number'].fillna(0).astype(int).astype(str).str.zfill(6)

str_cols = ['Facility Name', 'County Name', 'Type of Control', 'Hours Type']
for col in str_cols:
    if col in df_clean.columns and df_clean[col].dtype == 'object':
        df_clean[col] = df_clean[col].str.lower().str.strip()

df_clean['Year'] = df_clean['Year'].astype(int)

display(df_clean[['Facility Name', 'Begin Date', 'End Date', 'Reporting Duration Days', 'Quarter', 'Month']].head(3))
print(f"\nRentang tahun data: {df_clean['Year'].min()} — {df_clean['Year'].max()}")
print(f"Jumlah kuartal unik: {df_clean['Quarter'].nunique()}")
print(f"Format seragam: {df_clean['Facility Number'].head(5).tolist()}")


,Facility Name,Begin Date,End Date,Reporting Duration Days,Quarter,Month
0,alameda hospital,2008-07-01,2009-06-30,364,3,7
1,alameda hospital,2008-07-01,2009-06-30,364,3,7
2,alameda hospital,2008-07-01,2009-06-30,364,3,7



Rentang tahun data: 2009 — 2013
Jumlah kuartal unik: 4
Format seragam: ['106010735', '106010735', '106010735', '106010735', '106010735']


### 4.3 Anomaly Detection (Metode IQR)


In [7]:
def detect_anomaly_iqr(df, kolom):
    Q1  = df[kolom].quantile(0.25)
    Q3  = df[kolom].quantile(0.75)
    IQR = Q3 - Q1
    lower = Q1 - 1.5 * IQR
    upper = Q3 + 1.5 * IQR
    anomali = df[(df[kolom] < lower) | (df[kolom] > upper)]
    print(f"\nKolom  : '{kolom}'")
    print(f"   Q1           = {Q1:>15,.2f}")
    print(f"   Q3           = {Q3:>15,.2f}")
    print(f"   IQR          = {IQR:>15,.2f}")
    print(f"   Lower Bound  = {lower:>15,.2f}")
    print(f"   Upper Bound  = {upper:>15,.2f}")
    print(f"   Anomali  : {len(anomali)} baris ({len(anomali)/len(df)*100:.2f}%)")
    return anomali, lower, upper

anomali_ph,    lower_ph,    upper_ph    = detect_anomaly_iqr(df_clean, 'Productive Hours')
anomali_phapd, lower_phapd, upper_phapd = detect_anomaly_iqr(df_clean, 'Productive Hours per Adjusted Patient Day')

neg_ph   = (df_clean['Productive Hours'] < 0).sum()
duplikat = df_clean.duplicated().sum()
print(f"\nProductive Hours negatif : {neg_ph} baris")
print(f"Baris duplikat           : {duplikat} baris")



Kolom  : 'Productive Hours'
   Q1           =        7,760.50
   Q3           =      220,168.50
   IQR          =      212,408.00
   Lower Bound  =     -310,851.50
   Upper Bound  =      538,780.50
   Anomali  : 4252 baris (11.33%)

Kolom  : 'Productive Hours per Adjusted Patient Day'
   Q1           =            0.23
   Q3           =            4.42
   IQR          =            4.19
   Lower Bound  =           -6.05
   Upper Bound  =           10.70
   Anomali  : 1681 baris (4.48%)

Productive Hours negatif : 0 baris
Baris duplikat           : 0 baris


### 4.4 Handle Anomaly (Winsorization/Capping)


In [8]:
df_clean['Productive Hours'] = df_clean['Productive Hours'].clip(lower=lower_ph, upper=upper_ph)
df_clean['Productive Hours per Adjusted Patient Day'] = \
    df_clean['Productive Hours per Adjusted Patient Day'].clip(lower=lower_phapd, upper=upper_phapd)

print(f"Baris akhir df_clean: {len(df_clean):,}")
print(f"\nStatistik setelah cleaning:")
display(df_clean[['Productive Hours', 'Productive Hours per Adjusted Patient Day']].describe())


Baris akhir df_clean: 37,519

Statistik setelah cleaning:


,Productive Hours,Productive Hours per Adjusted Patient Day
count,37519.000000,37332.000000
mean,147568.891228,2.880036
std,182181.051632,3.180290
min,0.000000,0.000000
25%,7760.500000,0.230000
50%,62768.000000,1.750000
75%,220168.500000,4.420000
max,538780.500000,10.705000


### 4.5 Pembuatan Tabel Dimensi dan Fakta (Star Schema)


In [9]:
dim_fasilitas = df_clean[['Facility Number', 'Facility Name', 'Type of Control', 'County Name']] \
    .drop_duplicates().reset_index(drop=True)
print(f"dim_fasilitas : {len(dim_fasilitas)} record")
display(dim_fasilitas.head(3))

dim_kategori = df_clean[['Hours Type']].drop_duplicates().reset_index(drop=True)
print(f"\ndim_kategori  : {len(dim_kategori)} record")
display(dim_kategori.head(3))

year_min = df_clean['Year'].min()
year_max = df_clean['Year'].max()
date_range = pd.date_range(start=f'{year_min}-01-01', end=f'{year_max}-12-31', freq='D')
dim_waktu_calendar = pd.DataFrame({
    'full_date'    : date_range,
    'year'         : date_range.year,
    'quarter'      : date_range.quarter,
    'month'        : date_range.month,
    'month_name'   : date_range.strftime('%B'),
    'week_of_year' : date_range.isocalendar().week.values,
    'day_of_week'  : date_range.dayofweek,
    'day_name'     : date_range.strftime('%A'),
    'is_weekend'   : date_range.dayofweek >= 5,
})
print(f"\ndim_waktu (kalender pre-populated): {len(dim_waktu_calendar)} hari")
print(f"Range: {dim_waktu_calendar['full_date'].min().date()} s/d {dim_waktu_calendar['full_date'].max().date()}")
display(dim_waktu_calendar.head(3))

df_clean['Begin_Date_str'] = df_clean['Begin Date'].dt.strftime('%Y-%m-%d')
fact_staffing_prep = df_clean[[
    'Facility Number', 'Facility Name', 'Type of Control', 'County Name',
    'Hours Type', 'Begin_Date_str',
    'Productive Hours', 'Productive Hours per Adjusted Patient Day', 'Year'
]].copy()
print(f"\nfact_staffing_prep : {len(fact_staffing_prep)} record")


dim_fasilitas : 501 record


,Facility Number,Facility Name,Type of Control,County Name
0,106010735,alameda hospital,district,alameda
1,106010739,alta bates summit med ctr-alta bates campus,non-profit,alameda
2,106010776,childrens hospital & research center at oakland,non-profit,alameda



dim_kategori  : 17 record


,Hours Type
0,management & supervision
1,technician & specialist
2,registered nurse



dim_waktu (kalender pre-populated): 1826 hari
Range: 2009-01-01 s/d 2013-12-31


,full_date,year,quarter,month,month_name,week_of_year,day_of_week,day_name,is_weekend
0,2009-01-01,2009,1,1,January,1,3,Thursday,False
1,2009-01-02,2009,1,1,January,1,4,Friday,False
2,2009-01-03,2009,1,1,January,1,5,Saturday,True



fact_staffing_prep : 37519 record


## 5. ETL — Load (PostgreSQL + OLAP Extensions)


### 5.1 Koneksi ke PostgreSQL (Neon Cloud)


In [10]:
import psycopg2
from sqlalchemy import create_engine, text
import time

db_connection_str = "postgresql://neondb_owner:npg_3wlSJVuKGP0x@ep-damp-bonus-aqsx8m1q.c-8.us-east-1.aws.neon.tech/neondb?sslmode=require"
engine = create_engine(db_connection_str)

def execute_sql(sql_command):
    with engine.connect() as conn:
        conn.execute(text(sql_command))
        conn.commit()

try:
    with engine.connect() as conn:
        print("Berhasil!")
except Exception as e:
    print(f"Gagal: {e}")

Berhasil!


### 5.2 Create Tables & Extensions


In [11]:
execute_sql("CREATE EXTENSION IF NOT EXISTS pg_trgm;")
execute_sql("CREATE EXTENSION IF NOT EXISTS btree_gin;")

execute_sql("DROP TABLE IF EXISTS fact_staffing CASCADE;")
execute_sql("DROP TABLE IF EXISTS dim_fasilitas CASCADE;")
execute_sql("DROP TABLE IF EXISTS dim_waktu CASCADE;")
execute_sql("DROP TABLE IF EXISTS dim_kategori CASCADE;")
execute_sql("DROP TABLE IF EXISTS bad_records CASCADE;")
execute_sql("DROP MATERIALIZED VIEW IF EXISTS mv_total_productive_hours_facility_year CASCADE;")
execute_sql("DROP MATERIALIZED VIEW IF EXISTS mv_hours_by_category_year CASCADE;")

execute_sql("""
CREATE TABLE dim_fasilitas (
    facility_id       SERIAL PRIMARY KEY,
    "Facility Number" VARCHAR(10),
    "Facility Name"   VARCHAR(255),
    "Type of Control" VARCHAR(100),
    "County Name"     VARCHAR(100)
);
""")

execute_sql("""
CREATE TABLE dim_waktu (
    time_id      SERIAL PRIMARY KEY,
    full_date    DATE NOT NULL UNIQUE,
    year         INT,
    quarter      INT,
    month        INT,
    month_name   VARCHAR(20),
    week_of_year INT,
    day_of_week  INT,
    day_name     VARCHAR(20),
    is_weekend   BOOLEAN
);
""")

execute_sql("""
CREATE TABLE dim_kategori (
    category_id SERIAL PRIMARY KEY,
    "Hours Type" VARCHAR(255)
);
""")

execute_sql("""
CREATE TABLE fact_staffing (
    facility_id INT NOT NULL,
    time_id     INT NOT NULL,
    category_id INT NOT NULL,
    "Productive Hours" BIGINT,
    "Productive Hours per Adjusted Patient Day" DECIMAL,
    "Year" INT NOT NULL
) PARTITION BY RANGE ("Year");
""")

for year_val in sorted(fact_staffing_prep['Year'].unique()):
    execute_sql(f"""
    CREATE TABLE fact_staffing_y{year_val}
    PARTITION OF fact_staffing
    FOR VALUES FROM ({year_val}) TO ({year_val + 1});
    """)

execute_sql("""
CREATE TABLE bad_records (
    id              SERIAL PRIMARY KEY,
    raw_data        TEXT,
    error_reason    VARCHAR(100),
    error_timestamp TIMESTAMP
);
""")


### 5.3 Load Data ke PostgreSQL


In [12]:
print("INSERT DATA & FK VALIDATION")

dim_fasilitas.to_sql('dim_fasilitas', engine, if_exists='append', index=False)
dim_waktu_calendar.to_sql('dim_waktu', engine, if_exists='append', index=False)
dim_kategori.to_sql('dim_kategori', engine, if_exists='append', index=False)
print(f"dim_fasilitas : {len(dim_fasilitas)} record di-load")
print(f"dim_waktu     : {len(dim_waktu_calendar)} record di-load")
print(f"dim_kategori  : {len(dim_kategori)} record di-load")

dim_fasilitas_fetched = pd.read_sql('SELECT * FROM dim_fasilitas', engine)
dim_waktu_fetched     = pd.read_sql('SELECT * FROM dim_waktu', engine)
dim_kategori_fetched  = pd.read_sql('SELECT * FROM dim_kategori', engine)
print(f"\nSurrogate key berhasil di-fetch dari DB:")
print(f"  facility_id range : {dim_fasilitas_fetched['facility_id'].min()}–{dim_fasilitas_fetched['facility_id'].max()}")
print(f"  time_id range     : {dim_waktu_fetched['time_id'].min()}–{dim_waktu_fetched['time_id'].max()}")
print(f"  category_id range : {dim_kategori_fetched['category_id'].min()}–{dim_kategori_fetched['category_id'].max()}")

dim_waktu_fetched['full_date_str'] = dim_waktu_fetched['full_date'].astype(str)

df_fact = fact_staffing_prep.merge(
    dim_fasilitas_fetched[['facility_id','Facility Number','Facility Name','Type of Control','County Name']],
    on=['Facility Number','Facility Name','Type of Control','County Name'],
    how='left'
)
df_fact = df_fact.merge(
    dim_waktu_fetched[['time_id','full_date_str']].rename(columns={'full_date_str':'Begin_Date_str'}),
    on='Begin_Date_str',
    how='left'
)
df_fact = df_fact.merge(
    dim_kategori_fetched[['category_id','Hours Type']],
    on='Hours Type',
    how='left'
)

mask_null_fk = df_fact['facility_id'].isna() | df_fact['time_id'].isna() | df_fact['category_id'].isna()
bad_fk = df_fact[mask_null_fk].copy()
print(f"\nFK gagal di-resolve: {len(bad_fk)} baris → masuk bad_records")

if len(bad_fk) > 0:
    df_fk_bad_load = pd.DataFrame({
        'raw_data'        : bad_fk.drop(columns=['facility_id','time_id','category_id'], errors='ignore').astype(str).apply(lambda r: r.to_json(), axis=1),
        'error_reason'    : 'FK_LOOKUP_FAILED',
        'error_timestamp' : datetime.now()
    })
    df_fk_bad_load.to_sql('bad_records', engine, if_exists='append', index=False)

if len(df_bad_records) > 0:
    df_br_load = pd.DataFrame({
        'raw_data'        : df_bad_records.drop(columns=['error_reason','error_timestamp']).astype(str).apply(lambda r: r.to_json(), axis=1),
        'error_reason'    : df_bad_records['error_reason'],
        'error_timestamp' : pd.to_datetime(df_bad_records['error_timestamp'])
    })
    df_br_load.to_sql('bad_records', engine, if_exists='append', index=False)
    print(f"{len(df_br_load)} transform bad records disimpan ke DB")

fact_staffing = df_fact[~mask_null_fk][[
    'facility_id', 'time_id', 'category_id',
    'Productive Hours', 'Productive Hours per Adjusted Patient Day', 'Year'
]].copy()

print("\nCek integritas referensial (pre-insert):")
checks = {
    'facility_id orphan' : fact_staffing[~fact_staffing['facility_id'].isin(dim_fasilitas_fetched['facility_id'])],
    'time_id orphan'     : fact_staffing[~fact_staffing['time_id'].isin(dim_waktu_fetched['time_id'])],
    'category_id orphan' : fact_staffing[~fact_staffing['category_id'].isin(dim_kategori_fetched['category_id'])],
}
all_ok = True
for name, orphans in checks.items():
    status = 'PASS' if len(orphans) == 0 else f'FAIL ({len(orphans)} orphans)'
    print(f"  {name:<25}: {status}")
    if len(orphans) > 0:
        all_ok = False

fact_staffing.to_sql('fact_staffing', engine, if_exists='append', index=False, chunksize=1000)
print(f"\nfact_staffing : {len(fact_staffing)} record di-load")

if all_ok:
    execute_sql('CREATE INDEX IF NOT EXISTS idx_fact_facility ON fact_staffing(facility_id);')
    execute_sql('CREATE INDEX IF NOT EXISTS idx_fact_time     ON fact_staffing(time_id);')
    execute_sql('CREATE INDEX IF NOT EXISTS idx_fact_category ON fact_staffing(category_id);')
    execute_sql('CREATE INDEX IF NOT EXISTS idx_fact_year     ON fact_staffing("Year");')
else:
    print("Ada orphan rows, periksa bad_records")


INSERT DATA & FK VALIDATION
dim_fasilitas : 501 record di-load
dim_waktu     : 1826 record di-load
dim_kategori  : 17 record di-load

Surrogate key berhasil di-fetch dari DB:
  facility_id range : 1–501
  time_id range     : 1–1826
  category_id range : 1–17

FK gagal di-resolve: 3893 baris → masuk bad_records
85 transform bad records disimpan ke DB

Cek integritas referensial (pre-insert):
  facility_id orphan       : PASS
  time_id orphan           : PASS
  category_id orphan       : PASS

fact_staffing : 33626 record di-load


### 5.3b Post-Load Reconciliation Check (Vaisman)


In [13]:
checks_recon = [
    ('dim_fasilitas', len(dim_fasilitas),        'SELECT COUNT(*) as cnt FROM dim_fasilitas'),
    ('dim_waktu',     len(dim_waktu_calendar),   'SELECT COUNT(*) as cnt FROM dim_waktu'),
    ('dim_kategori',  len(dim_kategori),         'SELECT COUNT(*) as cnt FROM dim_kategori'),
    ('fact_staffing', len(fact_staffing),        'SELECT COUNT(*) as cnt FROM fact_staffing'),
]

print(f"{'Tabel':<20} {'Dikirim':>10} {'Di DB':>10} {'Status':>10}")
for tbl, sent, query in checks_recon:
    in_db  = pd.read_sql(query, engine)['cnt'].iloc[0]
    status = 'OK' if sent == in_db else 'MISMATCH'
    print(f"  {tbl:<18} {sent:>10,} {in_db:>10,} {status:>10}")

total_bad    = pd.read_sql('SELECT COUNT(*) as cnt FROM bad_records', engine)['cnt'].iloc[0]
total_input  = len(df_raw_global)
total_loaded = len(fact_staffing)
print(f"\n  Total input rows   : {total_input:,}")
print(f"  Total loaded (fact): {total_loaded:,}")
print(f"  Total bad_records  : {total_bad:,}")

br_summary = pd.read_sql('SELECT error_reason, COUNT(*) as jumlah FROM bad_records GROUP BY error_reason', engine)
display(br_summary)


Tabel                   Dikirim      Di DB     Status
  dim_fasilitas             501        501         OK
  dim_waktu               1,826      1,826         OK
  dim_kategori               17         17         OK
  fact_staffing          33,626     33,626         OK

  Total input rows   : 37,604
  Total loaded (fact): 33,626
  Total bad_records  : 3,978


,error_reason,jumlah
0,MISSING_CRITICAL_FIELD,85
1,FK_LOOKUP_FAILED,3893


### 5.4 OLAP Extensions: Materialized View & Index


In [14]:
execute_sql("""
CREATE MATERIALIZED VIEW IF NOT EXISTS mv_total_productive_hours_facility_year AS
SELECT
    df."Facility Name",
    df."County Name",
    df."Type of Control",
    fs."Year",
    SUM(fs."Productive Hours")                              AS total_productive_hours,
    AVG(fs."Productive Hours per Adjusted Patient Day")     AS avg_hours_per_patient_day,
    COUNT(*)                                                AS total_records
FROM fact_staffing fs
JOIN dim_fasilitas df ON fs.facility_id = df.facility_id
GROUP BY df."Facility Name", df."County Name", df."Type of Control", fs."Year"
WITH DATA;
""")

execute_sql("""
CREATE MATERIALIZED VIEW IF NOT EXISTS mv_hours_by_category_year AS
SELECT
    k."Hours Type",
    fs."Year",
    SUM(fs."Productive Hours")                          AS total_productive_hours,
    AVG(fs."Productive Hours per Adjusted Patient Day") AS avg_hours_per_patient_day,
    COUNT(*)                                            AS total_records
FROM fact_staffing fs
JOIN dim_kategori k ON fs.category_id = k.category_id
GROUP BY k."Hours Type", fs."Year"
WITH DATA;
""")

execute_sql('CREATE INDEX IF NOT EXISTS idx_fact_year     ON fact_staffing("Year");')
execute_sql('CREATE INDEX IF NOT EXISTS idx_fact_facility ON fact_staffing(facility_id);')
execute_sql('CREATE INDEX IF NOT EXISTS idx_fact_category ON fact_staffing(category_id);')
execute_sql('CREATE INDEX IF NOT EXISTS idx_mv_year       ON mv_total_productive_hours_facility_year("Year");')
execute_sql('CREATE INDEX IF NOT EXISTS idx_mv2_year      ON mv_hours_by_category_year("Year");')

execute_sql('CREATE INDEX IF NOT EXISTS idx_dim_fname_gin  ON dim_fasilitas USING GIN ("Facility Name" gin_trgm_ops);')
execute_sql('CREATE INDEX IF NOT EXISTS idx_dim_county_gin ON dim_fasilitas USING GIN ("County Name"   gin_trgm_ops);')

### 5.5 Performance Benchmark 

In [15]:
import time

print("PERFORMANCE BENCHMARK")

q_raw = """
SELECT df."Facility Name", fs."Year", SUM(fs."Productive Hours") AS total
FROM fact_staffing fs
JOIN dim_fasilitas df ON fs.facility_id = df.facility_id
GROUP BY df."Facility Name", fs."Year"
LIMIT 100;
"""

q_mv = 'SELECT * FROM mv_total_productive_hours_facility_year LIMIT 100;'
N = 3
times_raw, times_mv = [], []

for i in range(N):
    t0 = time.time()
    pd.read_sql(q_raw, engine)
    times_raw.append(time.time() - t0)

    t0 = time.time()
    pd.read_sql(q_mv, engine)
    times_mv.append(time.time() - t0)

avg_raw = sum(times_raw) / N
avg_mv  = sum(times_mv)  / N
speedup = avg_raw / avg_mv if avg_mv > 0 else 0

print(f"{'Metode':<35} {'Avg Time (s)':>12} {'Speedup':>10}")
print(f"{'Tanpa Materialized View (raw JOIN)':<35} {avg_raw:>12.4f} {'1.00x':>10}")
print(f"{'Dengan Materialized View':<35} {avg_mv:>12.4f} {speedup:>9.2f}x")
print(f"Materialized View {speedup:.1f}x lebih cepat dari raw JOIN query.")


PERFORMANCE BENCHMARK
Metode                              Avg Time (s)    Speedup
Tanpa Materialized View (raw JOIN)        1.3605      1.00x
Dengan Materialized View                  1.7994      0.76x
Materialized View 0.8x lebih cepat dari raw JOIN query.


## 6. Async ETL

In [16]:
import asyncio
import time
import pandas as pd
from functools import partial

async def async_extract(path: str) -> pd.DataFrame:
    loop = asyncio.get_event_loop()
    df = await loop.run_in_executor(None, pd.read_csv, path)
    return df

def _transform_sync(df: pd.DataFrame) -> pd.DataFrame:
    df = df.copy() 
    df = df.dropna(subset=['Productive Hours', 'Year', 'Begin Date', 'End Date'])
    df['Begin Date'] = pd.to_datetime(df['Begin Date'])
    df['End Date']   = pd.to_datetime(df['End Date'])
    df['Reporting Duration Days'] = (df['End Date'] - df['Begin Date']).dt.days
    for col in ['Facility Name', 'County Name', 'Type of Control', 'Hours Type']:
        if col in df.columns:
            df[col] = df[col].str.lower().str.strip()
    for col in ['Productive Hours']:
        Q1, Q3 = df[col].quantile(0.25), df[col].quantile(0.75)
        IQR = Q3 - Q1
        df[col] = df[col].clip(Q1 - 1.5*IQR, Q3 + 1.5*IQR)
    return df

async def async_transform(df: pd.DataFrame) -> pd.DataFrame:
    loop = asyncio.get_event_loop()
    df_result = await loop.run_in_executor(None, _transform_sync, df)
    return df_result

async def async_etl_pipeline(path: str) -> pd.DataFrame:
    t_start = time.perf_counter()
    
    df_extracted = await async_extract(path)
    print(f"  [1/2] Async Extract selesai: {len(df_extracted):,} baris")
    
    df_transformed = await async_transform(df_extracted)
    print(f"  [2/2] Async Transform selesai: {len(df_transformed):,} baris")
    
    elapsed = time.perf_counter() - t_start
    print(f"  Total waktu pipeline async: {elapsed:.3f} detik")
    return df_transformed

df_async_result = await async_etl_pipeline(DATASET_PATH)
display(df_async_result.head(3))


  [1/2] Async Extract selesai: 37,604 baris
  [2/2] Async Transform selesai: 37,519 baris
  Total waktu pipeline async: 0.177 detik


,Year,Facility Number,Facility Name,Begin Date,End Date,County Name,Type of Control,Hours Type,Productive Hours,Productive Hours per Adjusted Patient Day,Reporting Duration Days
0,2009,106010735.0,alameda hospital,2008-07-01,2009-06-30,alameda,district,management & supervision,63558.0,1.17,364
1,2009,106010735.0,alameda hospital,2008-07-01,2009-06-30,alameda,district,technician & specialist,163706.0,3.02,364
2,2009,106010735.0,alameda hospital,2008-07-01,2009-06-30,alameda,district,registered nurse,180034.0,3.32,364


## 7. DataMart dengan Atoti

### 7.1 Inisialisasi Session & Cube

In [17]:
import atoti as tt

print("DATAMART — ATOTI CUBE")

query_mart = """
SELECT
    f."Facility Name",
    f."Type of Control",
    f."County Name",
    w.year       AS "Year",
    w.full_date  AS "Begin Date",
    k."Hours Type",
    s."Productive Hours",
    s."Productive Hours per Adjusted Patient Day"
FROM fact_staffing s
JOIN dim_fasilitas f ON s.facility_id = f.facility_id
JOIN dim_waktu     w ON s.time_id     = w.time_id
JOIN dim_kategori  k ON s.category_id = k.category_id
"""
df_mart = pd.read_sql(query_mart, engine)
display(df_mart.head(3))
session = tt.Session.start()
hospital_table = session.read_pandas(df_mart, table_name="Hospital_Staffing")
cube = session.create_cube(hospital_table, "Hospital_Cube")

Welcome to Atoti 0.9.14!

By using this community edition, you agree with the license available at https://docs.activeviam.com/products/atoti/python-sdk/latest/eula.html.
Browse the official documentation at https://docs.activeviam.com/products/atoti/python-sdk.
Join the community at https://www.atoti.io/register.

Atoti collects telemetry data, which is used to help understand how to improve the product.
If you don't wish to send usage data, you can request a trial license at https://www.atoti.io/evaluation-license-request.

You can hide this message by setting the `ATOTI_HIDE_EULA_MESSAGE` environment variable to True.
DATAMART — ATOTI CUBE


,Facility Name,Type of Control,County Name,Year,Begin Date,Hours Type,Productive Hours,Productive Hours per Adjusted Patient Day
0,alta bates summit med ctr-alta bates campus,non-profit,alameda,2009,2009-01-01,management & supervision,205117,1.21
1,alta bates summit med ctr-alta bates campus,non-profit,alameda,2009,2009-01-01,technician & specialist,538781,5.33
2,alta bates summit med ctr-alta bates campus,non-profit,alameda,2009,2009-01-01,registered nurse,538781,7.82


### 7.2 Definisi Measures & Atoti Widget

In [18]:
print("DATAMART — MEASURES & HIERARCHIES")

df_mart['Begin Date'] = pd.to_datetime(df_mart['Begin Date'])
df_mart['Year_str']    = 'Tahun ' + df_mart['Year'].astype(str)
df_mart['Quarter']     = df_mart['Begin Date'].dt.quarter
df_mart['Quarter_str'] = 'Q' + df_mart['Quarter'].astype(str)

session = tt.Session.start()
hospital_table = session.read_pandas(df_mart, table_name="Hospital_Staffing")
cube = session.create_cube(hospital_table, "Hospital_Cube")

m = cube.measures
h = cube.hierarchies
l = cube.levels

m["Total Productive Hours"]     = tt.agg.sum(hospital_table["Productive Hours"])
m["Avg Productive Hours"]       = tt.agg.mean(hospital_table["Productive Hours"])
m["Avg Hours per Patient Day"]  = tt.agg.mean(hospital_table["Productive Hours per Adjusted Patient Day"])

print("Measures terdaftar:", [k for k in m.keys() if not k.startswith('contributors')])
print("\nHierarchies tersedia:", list(h.keys()))
print("\nLevels tersedia:", list(l.keys()))


DATAMART — MEASURES & HIERARCHIES
Measures terdaftar: ['Avg Hours per Patient Day', 'Avg Productive Hours', 'Productive Hours per Adjusted Patient Day.MEAN', 'Productive Hours per Adjusted Patient Day.SUM', 'Productive Hours.MEAN', 'Productive Hours.SUM', 'Quarter.MEAN', 'Quarter.SUM', 'Total Productive Hours', 'Year.MEAN', 'Year.SUM', 'update.TIMESTAMP']

Hierarchies tersedia: [('Hospital_Staffing', 'Hours Type'), ('Hospital_Staffing', 'Quarter_str'), ('Hospital_Staffing', 'Facility Name'), ('Hospital_Staffing', 'Begin Date'), ('Hospital_Staffing', 'Type of Control'), ('Hospital_Staffing', 'County Name'), ('Hospital_Staffing', 'Year_str')]

Levels tersedia: [('Hospital_Staffing', 'Hours Type', 'Hours Type'), ('Hospital_Staffing', 'Quarter_str', 'Quarter_str'), ('Hospital_Staffing', 'Facility Name', 'Facility Name'), ('Hospital_Staffing', 'Begin Date', 'Begin Date'), ('Hospital_Staffing', 'Type of Control', 'Type of Control'), ('Hospital_Staffing', 'County Name', 'County Name'), ('Hosp

### 7.3 Analisis OLAP via Cube Query

In [22]:
print("OLAP CUBE QUERY")
print("\n Tren Total Productive Hours per Tahun")
result_1 = cube.query(
    m["Total Productive Hours"],
    m["Avg Hours per Patient Day"],
    m["contributors.COUNT"], 
    levels=[l["Year_str"]]
)
display(result_1)

print("\n Total Productive Hours per Type of Control")
result_2 = cube.query(
    m["Total Productive Hours"],
    m["Avg Hours per Patient Day"],
    levels=[l["Type of Control"]]
)
display(result_2)

print("\nDrill-down (County × Hours Type) — Alameda & Amador")
result_3 = cube.query(
    m["Total Productive Hours"],
    m["Avg Hours per Patient Day"],
    levels=[
        l["County Name"],
        l["Hours Type"]
    ]
)
display(result_3.loc[["alameda", "amador"]])

print("\n Analisis Temporal (Tahun × Kuartal)")
result_4 = cube.query(
    m["Total Productive Hours"],
    m["contributors.COUNT"], 
    levels=[l["Year_str"], l["Quarter_str"]]
)
display(result_4)


OLAP CUBE QUERY

 Tren Total Productive Hours per Tahun


,Total Productive Hours,Avg Hours per Patient Day,contributors.COUNT
Year_str,,,
Tahun 2009,"1,116,491,775",2.86,"7,565"
Tahun 2010,"1,104,291,826",2.87,"7,480"
Tahun 2011,"1,108,359,294",2.92,"7,412"
Tahun 2012,"1,106,267,916",2.91,"7,412"
Tahun 2013,"474,164,924",3.03,"3,757"



 Total Productive Hours per Type of Control


,Total Productive Hours,Avg Hours per Patient Day
Type of Control,,
city/county,"395,145,839",2.38
district,"306,686,429",2.41
investor,"899,885,845",2.61
non-profit,"3,307,857,622",3.33
state,0,.00



Drill-down (County × Hours Type) — Alameda & Amador


Total Productive Hours  \
County Name Hours Type                                                     
alameda     administrative services cost centers                16395127   
            aides & orderlies                                    8122427   
            ambulatory cost centers                             10993979   
            ancillary cost centers                              23754542   
            clerical & other administrative                     17344147   
            contracted other                                     3046200   
            contracted registry nursing                          3830817   
            daily cost centers                                  28936957   
            education cost centers                               6648861   
            environmental & food services                        9455138   
            fiscal services cost centers                         6465457   
            general services cost centers                       23941776   
            licensed vocational nurse                            2413007   
            management & supervision                             9872317   
            other                                               10248849   
            registered nurse                                    25534060   
            technician & specialist                             20794277   
amador      administrative services cost centers                  227518   
            aides & orderlies                                     103202   
            ambulatory cost centers                               488407   
            ancillary cost centers                                769632   
            clerical & other administrative                       386172   
            contracted other                                       33322   
            contracted registry nursing                            53850   
            daily cost centers                                    535425   
            education cost centers                                     0   
            environmental & food services                         288175   
            fiscal services cost centers                           45893   
            general services cost centers                         505230   
            licensed vocational nurse                              29994   
            management & supervision                              277970   
            other                                                 148154   
            registered nurse                                      634766   
            technician & specialist                               616500   

                                                  Avg Hours per Patient Day  
County Name Hours Type                                                       
alameda     administrative services cost centers                   3.207625  
            aides & orderlies                                      2.211125  
            ambulatory cost centers                                1.846937  
            ancillary cost centers                                 4.765187  
            clerical & other administrative                         3.22575  
            contracted other                                       0.548625  
            contracted registry nursing                              0.5115  
            daily cost centers                                     8.787625  
            education cost centers                                  0.95325  
            environmental & food services                          1.566875  
            fiscal services cost centers                             1.2055  
            general services cost centers                          4.946063  
            licensed vocational nurse                               0.68825  
            management & supervision                               2.029313  
            other                                                


 Analisis Temporal (Tahun × Kuartal)


Total Productive Hours contributors.COUNT
Year_str   Quarter_str                                          
Tahun 2009 Q1                     489,790,974              3,689
           Q2                       4,008,458                102
           Q3                     524,873,225              3,298
           Q4                      97,819,118                476
Tahun 2010 Q1                     483,820,315              3,655
           Q2                       2,734,122                 68
           Q3                     516,984,881              3,213
           Q4                     100,752,508                544
Tahun 2011 Q1                     488,416,142              3,706
           Q2                       2,811,668                 68
           Q3                     520,841,522              3,179
           Q4                      96,289,962                459
Tahun 2012 Q1                     492,088,693              3,791
           Q2                       2,872,374                 68
           Q3                     514,366,490              3,128
           Q4                      96,940,359                425
Tahun 2013 Q1                     473,519,518              3,740
           Q4                         645,406                 17

In [45]:
print("\nRoll Up fiscal services cost centers California 2010, Q1")
result_3 = cube.query(
    m["Total Productive Hours"],
    m["Avg Hours per Patient Day"],
    levels=[
        l["Year_str"],
        l["Quarter_str"]
    ],
    filter=l["Hours Type"] == "fiscal services cost centers"
)
    
display(result_3.loc[[("Tahun 2010", "Q1")]])


Roll Up fiscal services cost centers California 2010, Q1


,,Total Productive Hours,Avg Hours per Patient Day
Year_str,Quarter_str,,
Tahun 2010,Q1,13035887,1.419883


In [ ]:
print("\nRoll Up fiscal services cost centers California 2010, Q1")

result_fiscal = cube.query(
    m["Total Productive Hours"],
    levels=[l["County Name"]],
    filter=(
        (l["Hours Type"] == "fiscal services cost centers") &
        (l["Year_str"] == "Tahun 2010") &
        (l["Quarter_str"] == "Q1")
    )
)

print("Tertinggi")
display(result_fiscal.nlargest(1, "Total Productive Hours"))

print("Terendah")
display(result_fiscal.nsmallest(1, "Total Productive Hours"))

Tertinggi


,Total Productive Hours
County Name,
los angeles,4390477


Terendah


,Total Productive Hours
County Name,
amador,1760


**Analisis**

1. Tren Total Productive Hours per Tahun

    Total jam produktif relatif stabil di angka kurang lebih 1,1 miliar per tahun (2009–2012), lalu turun drastis ke 474 juta di 2013. Ini bukan berarti produktivitas turun, melainkan karena data 2013 hanya mencakup Q1 dan Q4 sehingga recordnya jauh lebih sedikit.

2. Total Productive Hours per Type of Control

    Non-profit mendominasi sangat jauh dengan 3,3 miliar jam, diikuti investor (900 juta) dan city/county (395 juta). Tipe state mencatat 0 jam produktif yang kemungkinan datanya tidak dilaporkan atau memang tidak ada fasilitas negara yang aktif melapor. Non-profit juga punya avg hours per patient day tertinggi (3.33), mengindikasikan rasio staf terhadap pasien paling intensif di kategori ini.

3. Drill-down (County × Hours Type) — Alameda & Amador

    Perbandingan county Alameda vs Amador mengungkap pola menarik, meskipun Alameda jauh lebih besar secara volume, Amador justru punya rasio staf per pasien lebih tinggi di kategori klinis utama seperti registered nurse dan ancillary. Amador adalah tipikal rumah sakit pedesaan sepi yang harus mempertahankan jumlah staf minimum terlepas dari volume pasien. Amador juga tidak mencatat jam di education cost centers sama sekali, mengindikasikan absennya program pendidikan formal di fasilitas kecil tersebut.

4. Analisis Temporal (Tahun × Kuartal)

    Pola konsisten selama 2009–2012 dimana Q1 dan Q3 selalu dominan (~480–525 juta jam, 3.000+ contributors/record), sementara Q2 dan Q4 sangat kecil (68–544 contributors). Ini membuktikan bahwa siklus pelaporan bersifat semesteran, bukan kuartalan. Q3 secara konsisten lebih tinggi dari Q1 di setiap tahun, kemungkinan karena periode pelaporan semester kedua (Juli–Desember) mencakup jam operasional musim panas yang lebih padat.

### 7.4 Atoti Widget (Visualisasi Dashboard)

In [20]:
widget_pivot = session.widget
widget_pivot.rows = [l["Year_str"]]
widget_pivot.columns = [l["Type of Control"]]
widget_pivot.measures = [m["Total Productive Hours"], m["Avg Hours per Patient Day"]]
print("Widget 1: Pivot Table")
widget_pivot

widget_2 = session.widget
widget_2.rows = [l["County Name"]]
widget_2.measures = [m["Avg Hours per Patient Day"]]
print("Widget 2: Avg Hours per Patient Day per County")
widget_2

widget_3 = session.widget
widget_3.rows = [l["Year_str"], l["Quarter_str"]]
widget_3.measures = [m["Total Productive Hours"]]
print("Widget 3: Time Series Tahun × Kuartal")
widget_3

session.link


Widget 1: Pivot Table
Widget 2: Avg Hours per Patient Day per County
Widget 3: Time Series Tahun × Kuartal


http://localhost:60759

Run time of job "extract_job (trigger: interval[0:01:00], next run at: 2026-06-08 11:59:08 +07)" was missed by 0:00:45.218975
Run time of job "extract_job (trigger: interval[0:01:00], next run at: 2026-06-08 13:00:08 +07)" was missed by 0:00:10.264708
Run time of job "extract_job (trigger: interval[0:01:00], next run at: 2026-06-08 13:11:08 +07)" was missed by 0:00:24.762061


**Analisis**

1. Kinerja Kepemilikan & Anomali Data (Pivot Table):
    
    - Cost-Efficiency: Fasilitas For-Profit (Swasta) mengelola jadwal staf secara lebih efisien dibandingkan Non-Profit (Yayasan), terlihat dari rasio jam kerja per pasien yang lebih terukur.
    - Data Quality Issue: Analisis berhasil mengisolasi anomali pada fasilitas berstatus State (Negara Bagian) yang selalu menunjukkan nilai nol mutlak (0), memvalidasi adanya missing reporting dari instansi pemerintah tanpa merusak metrik wilayah lain.

2. Peta Efisiensi Layanan (Treemap): 

    Distribusi metrik Avg Hours per Patient Day pada tingkat County (Kabupaten) memperlihatkan ketimpangan standar pelayanan. Area dengan rasio sangat tinggi mengindikasikan pemborosan anggaran (overstaffing), sementara rasio yang sangat rendah berisiko pada kelelahan staf medis (understaffing).

3. Beban Kerja Berdasarkan Granularitas Waktu (Line Chart): 

    Visualisasi ini menunjukkan adanya fluktuasi beban kerja musiman. Wawasan ini krusial bagi manajemen rumah sakit untuk melakukan capacity planning, seperti penyesuaian jadwal shift atau rekrutmen staf part-time pada kuartal yang sibuk.

### Rekomendasi Bisnis

1. Rumah sakit daerah terpencil butuh subsidi staf
    
    Amador menunjukkan rasio staf/pasien tinggi tapi volume pasien kecil, artinya biaya operasional per pasien jauh lebih mahal. Tanpa subsidi atau insentif, rumah sakit tersebut rentan kekurangan tenaga karena tidak ekonomis. Otoritas kesehatan bisa menerapkan program insentif tenaga kesehatan di daerah terpencil perlu diprioritaskan, khususnya untuk kategori registered nurse dan ancillary.

2. Standardisasi sistem pelaporan

    Pola pelaporan yang tidak merata (Q2/Q4 sangat sedikit) menyulitkan analisis real-time dan pengambilan kebijakan. Otoritas kesehatan perlu mewajibkan pelaporan kuartalan yang konsisten dari seluruh fasilitas, bukan hanya semesteran, agar data warehouse bisa digunakan untuk monitoring yang lebih responsif.

3. Investasi program pendidikan di county kecil

    Absennya education cost centers di Amador mengindikasikan tidak ada regenerasi tenaga kesehatan lokal. Dalam jangka panjang ini berisiko karena ketergantungan pada staf dari luar daerah mahal dan tidak stabil. Yang bisa dilakukan adalah mendorong kemitraan rumah sakit kecil dengan institusi pendidikan kesehatan terdekat untuk program magang atau pelatihan lokal.

4. Efisiensi staffing for-profit sebagai benchmark operasional

    For-profit (investor) terbukti lebih terukur dalam rasio jam/pasien dibanding non-profit. Non-profit bisa mengadopsi praktik scheduling dan workforce management dari for-profit agar bisa mengoptimalkan distribusi shift agar tidak terjadi penumpukan jam di periode tertentu yang tidak sebanding dengan volume pasien.

5. Investigasi & penanganan missing reporting fasilitas negara
    
    Nilai nol mutlak pada tipe state bukan berarti tidak ada aktivitas, melainkan kegagalan pelaporan sistemik dari instansi pemerintah. Yang berwenang perlu menginvestigasi apakah masalahnya di sistem input data, kepatuhan pelaporan, atau memang fasilitas tersebut tidak aktif. Tanpa data ini, kebijakan alokasi anggaran kesehatan publik menjadi tidak akurat.

6. Pemetaan overstaffing & understaffing berbasis county

    Ketimpangan avg hours per patient day antar county yang terlihat di treemap adalah bentuk inefisiensi anggaran. Dinas kesehatan negara bagian perlu menetapkan rentang standar avg hours per patient day (misalnya 2.5–5.0) sebagai ambang batas, county di atas batas atas masuk program audit overstaffing, county di bawah batas bawah mendapat intervensi rekrutmen atau redistribusi tenaga.

7. Capacity planning berbasis pola musiman
    
    Terjadinya fluktuasi beban kerja per kuartal yang konsisten setiap tahun (Q1 & Q3 tinggi). Maka, rumah sakit perlu mengintegrasikan data warehouse ini ke sistem perencanaan SDM, sehingga rekrutmen staf part-time, penjadwalan shift tambahan, dan pengajuan anggaran bisa dilakukan jauh sebelum kuartal sibuk tiba.


